Enable variable sampling period

churanchen opened the pull request:
We have a DynamicConfigLoader. It is initialized with a channel and a time duration. The channel is used by the controller (also known as the pusher) to shut the DynamicConfigLoader down. The time duration is how often we check to see if the config has changed.

This has been implemented for stdout and prometheus.

We changed the sampling period by stopping and creating a new Ticker for the controller.

You can test by running `go run main.go` under example/dynamicconfig. The sampling period is initially a minute, but after roughly ten seconds, metrics will be outputted every two seconds

vmingchen commented on exporters/dynamicconfigloader/dynamicconfigloader.go:
```diff
@@ -0,0 +1,74 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dynamicconfigloader
```
I feel "loader" sounds like a one-time tool that load config from some backends. A "watcher" is probably closer to the meaning of watch config changes. This module can define

MetricConfig struct { Period time.Duration; }
ConfigWatcher interface {
OnInitialConfig(MetricConfig);
OnUpdatedConfig(MetricConfig);
}

Register(ConfigWatcher watcher)
Unregister(ConfigWatcher watcher)

Then, in controller/push.go, it can implement a config watcher, register the watch on construction, and unregister on destruction.

We can call this module confignotifier, it will have a goroutine to periodically check config changes. If there is a change, it should call the OnUpdatedConfig() for all registered listeners.

vmingchen replied:
Also, we will add parts to periodically check config changes, so we can simply name the directory "dynamicconfig".

vmingchen commented on exporters/dynamicconfigloader/dynamicconfigloader.go:
```diff
@@ -0,0 +1,74 @@
+	"fmt"
+	"time"
+
+	"go.opentelemetry.io/otel/sdk/metric/controller/push"
```
As a config module, it should not depend on the controller/push module, the dependency should be the other way around, right?

vmingchen commented on exporters/metric/prometheus/prometheus.go:
```diff
@@ -162,8 +163,11 @@ func NewExportPipeline(config Config, period time.Duration) (*push.Controller, h
 	//
 	// Gauges (or LastValues) and Summaries are an exception to this and have different behaviors.
 	integrator := integrator.New(selector, true)
-	pusher := push.New(integrator, exporter, period)
```
With a watcher-like API, we should be able to simply replace "period" with a config notifier.

vmingchen commented on sdk/metric/controller/push/push.go:
```diff
@@ -28,20 +28,21 @@ import (
-	exporter     export.Exporter
-	wg           sync.WaitGroup
-	ch           chan struct{}
-	period       time.Duration
```
"period" should not be needed any more.

vmingchen replied:
Also, is the formatting done automatically by "gofmt"? Please check opentelemetry's repository to see if it has its own formatting config, if yes, make sure we are using that.

churanchen replied:
Good catch!

vmingchen commented on sdk/metric/controller/push/push.go:
```diff
@@ -153,14 +156,36 @@ func (c *Controller) Stop() {
+	close(c.configLoaderCh)
+}
+
+func (c *Controller) RestartTicker(period time.Duration) {
```
This pretty much becomes the implementation of the config watcher's OnUpdatedConfig method.

vmingchen commented on sdk/metric/controller/push/push.go:
```diff
@@ -73,24 +74,26 @@ var _ Ticker = realTicker{}
 	}
 
 	impl := sdk.NewAccumulator(integrator, sdk.WithErrorHandler(c.ErrorHandler))
 	return &Controller{
```
We can implement a ConfigWatcher, register it with the notifier, and pass it to controller.

churanchen replied:
Let me know if I implemented this in the way you were thinking.

vmingchen commented on sdk/metric/controller/push/push.go:
```diff
@@ -73,24 +74,26 @@ var _ Ticker = realTicker{}
+		integrator:   	integrator,
+		exporter:     	exporter,
+		ch:           	make(chan bool),
+		period:       	period,
```
The initial period should be read from OnInitialConfig()

churanchen replied:
Let me know if I implemented this in the way you were thinking.

churanchen replied:
As part of my implementation, I set the period to a minute for safety reasons if `OnInitialConfig` is never called. Again, let met know if you had a different implementation in mind!

vmingchen commented on sdk/metric/controller/push/push.go:
```diff
@@ -153,14 +156,36 @@ func (c *Controller) Stop() {
 	for {
 		select {
-		case <-ch:
+		// If signal receives 'true', break to check the new ticker
```
Unregister the watcher here.

It is okay for the notifier to outlive the controller, so we don't need to bother with closing the notifier here.

churanchen replied:
I unregister the watcher in the controller's `stop` method, if that's okay

vmingchen commented on exporters/dynamicconfig/confignotifier.go:
```diff
@@ -0,0 +1,99 @@
+
+type ConfigNotifier struct {
+	checkFrequency time.Duration
+	config         *MetricConfig
```
this needs to be protected by a mutex because it can be updated by the "checkChanges()" goroutines while being read by Register()

vmingchen replied:
Similarly for "subscribed" in case there are concurrent calls of Register() and Unregister()

wtong98 commented on exporters/dynamicconfig/confignotifier.go:
```diff
@@ -0,0 +1,99 @@
+	Period time.Duration
+}
+
+func (config *MetricConfig) equals(otherConfig *MetricConfig) bool {
+	if config == nil && otherConfig == nil {
+		return true
+	}
+
+	if config == nil || otherConfig == nil {
+		return false
+	}
+
+	configString := fmt.Sprintf("%+v", *config)
+	otherConfigString := fmt.Sprintf("%+v", *otherConfig)
+
+	if configString == otherConfigString {
+		return true
+	}
+	return false
+}
```
Could we simplify this logic to:
```go
// make config strings

isEqual := configString == otherConfigString
return isEqual
```
If both `configString` and `otherConfigString` are `nil`, would `fmt.Sprintf` still yield the same string for both at the end?

wtong98 replied:
What are your thoughts on `cmp.Equal`, which may be more rigorous than converting to strings?

churanchen replied:
I separate the config equality logic into it's own function because this entire function body is probably going to end up replaced in the end. The reason we can do a simple string comparison right now is because the config only has one field.  In the future, when we have a better idea of what a config would look like, this function body will probably do some sort of deep comparison.

I'm not sure about `cmp.Equal`. Apparently it's not very performant and mostly meant for testing, which might an issue if the config is sufficiently big, and we check a remote service sufficiently often (https://godoc.org/github.com/google/go-cmp/cmp).

wtong98 commented on exporters/dynamicconfig/confignotifier.go:
```diff
@@ -0,0 +1,99 @@
+		subscribed:     make(map[ConfigWatcher]bool),
+	}
+
+	go configNotifier.checkChanges()
```
It seems like this function does two things: create a new *ConfigNotifier and start it. What do you think about separating these two tasks into two separate functions? Maybe a separate `func (cn *ConfigNotifier) Start()` that starts the process?

wtong98 commented on sdk/metric/controller/push/push.go:
```diff
@@ -73,24 +75,26 @@ var _ Ticker = realTicker{}
 // configuration options to configure an SDK with periodic collection.
 // The integrator itself is configured with the aggregation selector policy.
-func New(integrator export.Integrator, exporter export.Exporter, period time.Duration, opts ...Option) *Controller {
+func New(integrator export.Integrator, exporter export.Exporter, configNotifier *notifier.ConfigNotifier,
+	opts ...Option) *Controller {
```
Instead of introducing a new field for the configNotifier, what do you think about bundling it in with the Options logic instead? That way, maybe we can maintain compatibility with older code and don't have to worry about breaking existing projects that use `push.New()`?

churanchen replied:
Thanks for the look over!

Your first point got me thinking whether ConfigNotifier should be a mandatory field in the Controller. I talked it over with @vmingchen, and we decided that it should be mandatory, with the option for a config that is constant that we never change. This means we can't fold it into the Options logic, but in any case, thanks for bringing this up! It brought up an important point.

As to your second point, I think this might be problematic in that we'd have to register the notifier with every single different exporter that ends up getting implemented for open telemetry, and we want to try to keep as much as the logic inside the controller and dynamicconfig modules as we can.

churanchen commented:
These last few commits do a few things:
-I resolved merge conflicts
-I added unit tests for configNotifier
-I added a `stop` mechanism for configNotifier. Even if it's not strictly necessary, might be useful later

vmingchen commented on exporters/dynamicconfig/confignotifier.go:
```diff
@@ -55,45 +68,114 @@ type ConfigWatcher interface {
+	defer n.lock.Unlock()
+
+	n.subscribed[watcher] = true
+	watcher.OnInitialConfig(n.config)
```
Let's don't hold the lock while calling "watcher.OnInitialConfig" because who knows how long the watcher is going to fiddle with the config.

vmingchen commented on exporters/dynamicconfig/confignotifier.go:
```diff
@@ -55,45 +68,114 @@ type ConfigWatcher interface {
+			if !n.config.equals(newConfig) {
+				n.config = newConfig
+
+				for watcher := range n.subscribed {
```
Let's do not hold the lock here as well. It means we may need to copy the pointers for all watchers, but that is okay.

This frees us from lock starvation. A more subtle bug is deadlock if there were some lock dependency among the watchers.

vmingchen commented on exporters/dynamicconfig/confignotifier.go:
```diff
@@ -35,17 +39,26 @@ func (config *MetricConfig) equals(otherConfig *MetricConfig) bool {
+// We don't do anything with the configHost string right now
+// TIMES_READ_TEST is for testing purposes, to mimic a dynamic service
+// The second time we call readConfig, the sampling period changes
+var TIMES_CONFIG_READ_TEST = 0
```
Let's add a TODO to fix this later. A good approach is to passing a mock of config service.

churanchen replied:
Is the TODO on [https://github.com/vmingchen/opentelemetry-go/blob/55cd0019794079ee2efbf498a1df25586daccecd/exporters/dynamicconfig/confignotifier.go#L45](line 45) enough, or should I add another TODO?

vmingchen commented on exporters/dynamicconfig/confignotifier.go:
```diff
@@ -55,45 +68,114 @@ type ConfigWatcher interface {
 }
 
 type ConfigNotifier struct {
```
The list is getting longer. Let's add a comment for each member.

vmingchen commented on exporters/dynamicconfig/confignotifier.go:
```diff
@@ -55,45 +68,114 @@ type ConfigWatcher interface {
+	}
+
+	n.ticker = n.clock.Ticker(n.checkFrequency)
+	n.wg.Add(1)
```
Why is this needed? We allow a notifier to be started and stopped multiple times? If yes, why?

churanchen replied:
I might have misunderstood your comment, so feel free to clarify! It's not necessary for a notifier to be able to stop and start.

The wg means that when we run ConfigNotifier.Stop(), it will wait for checkChanges to return before we Stop the entire thing. I added this because it seems more "correct" for the Stop method to stop everything associated with the ConfigNotifier, rather than letting a loose goroutine to continue.

If you mean why do we have a Stop() method at all, I decided to add it because it seems to be bad form for a goroutine to not have a stop condition, and it sounds plausible that at some point, we might want a stop mechanism. Let me know if you disagree!

vmingchen commented on exporters/dynamicconfig/confignotifier.go:
```diff
@@ -167,15 +178,23 @@ func (n *ConfigNotifier) checkChanges(ch chan struct{}) {
+				}
+				n.lock.Unlock()
+
 				for watcher := range n.subscribed {
```
Iterate the copy instead.

vmingchen commented on exporters/dynamicconfig/confignotifier.go:
```diff
@@ -167,15 +178,23 @@ func (n *ConfigNotifier) checkChanges(ch chan struct{}) {
+				i := 0
+				for watcher := range n.subscribed {
+					subscribed_copy[i] = watcher
+					i++
```
for i, watcher in enumerate(xxx)?

churanchen replied:
I don't think there was a better way to do it, in terms of getting keys from a map.

It is cleaner to just copy over a new map, though copying just the keys into a list might be slightly more performant. Let me know your thoughts.

vmingchen commented on exporters/dynamicconfig/confignotifier.go:
```diff
@@ -145,9 +156,9 @@ func (n *ConfigNotifier) Stop() {
 	n.subscribed[watcher] = true
+	n.lock.Unlock()
+
 	watcher.OnInitialConfig(n.config)
```
Make a copy of n.config in the critical section and use it outside the critical section.

vmingchen replied:
TODO: Do we have a go equivalent of C++'s thread-safety analysis?

https://clang.llvm.org/docs/ThreadSafetyAnalysis.html

vmingchen replied:
Go tsan.

churanchen replied:
By go tsan, do you mean this? https://github.com/google/sanitizers

So open telemetry actually has tsan enabled by default in its makefile! `make test` also looks for race conditions

vmingchen commented on exporters/dynamicconfig/confignotifier.go:
```diff
@@ -0,0 +1,200 @@
+// configHost is the IP address of the config service. Can be set to ""
+// subscribed is a map of all configNotifier's current subscribers
+// ticker controls when we check the config service for a new config
+// wg is used to wait for the config-checking routine to return before stopping
```
Is this the go way of adding comment for struct field?

churanchen replied:
Doing it by comment by comment seems more widely used, so changed it!

churanchen commented:
Resolved another merge conflict, and made everything optional.

vmingchen commented on exporters/dynamicconfig/confignotifier.go:
```diff
@@ -0,0 +1,207 @@
+	config_copy := *n.config
+	n.lock.Unlock()
+
+	watcher.OnInitialConfig(&config_copy)
```
Thinking more about this I think it may be problematic if we do this like this. The following sequence of events is possible (in chronicle order):

1. Watcher (W1) register an initial config handle.
2. The watcher W1 is added, but "watcher.OnInitialConfig" is not called because the OS decides to schedule something else.
3. A new config config come.
4. The new config is applied to W1 in checkChanges() with watcher.OnUpdatedConfig().
5. The Register() is re-scheduled by the OS and resumes its execution, so "watcher.OnInitialConfig(&config_copy)" is called.

So, we end up overwriting the new updated config and incorrectly using the old initial config as the most up-to-date config.

churanchen commented:
Addressed comments. I did not remove the lock from my test code, as it is needed to prevent race conditions. This is because testLock is used to simulate a watcher's lock, rather than the notifier's lock.

churanchen commented on exporters/dynamicconfig/confignotifier_test.go:
```diff
@@ -0,0 +1,120 @@
+// testLock is to prevent race conditions in test code
+// testVar is used to verify OnInitialConfig and OnUpdatedConfig are called
+type testWatcher struct {
+	testLock sync.Mutex
```
Yep. Even if OnInitialConfig() and OnUpdateConfig can't be called concurrently, we could try to read testVar at the same time as one of them updating testVar.
